File: src/src_platform_swing/fmg/swing/draw/mosaic/CellPaint.java
package fmg.swing.draw.mosaic;

import fmg.core.mosaic.cells.BaseCell;
import fmg.core.mosaic.draw.ICellPaint;
import fmg.core.mosaic.draw.IPaintable;

/**
 * Class for drawing cell
 *
 * @param <TPaintable> see {@link IPaintable}
 * @param <TImage> SWING specific image: {@link java.awt.Image} or {@link javax.swing.Icon}
 */
public abstract class CellPaint<TPaintable extends IPaintable, TImage> implements ICellPaint<TPaintable, TImage, PaintSwingContext<TImage>> {

   @Override
   public abstract void paint(BaseCell cell, TPaintable p, PaintSwingContext<TImage> paintContext);

   protected abstract void paintBorder(BaseCell cell, TPaintable p, PaintSwingContext<TImage> paintContext);

   /** draw border lines */
   protected abstract void paintBorderLines(BaseCell cell, TPaintable p, PaintSwingContext<TImage> paintContext);

   protected abstract void paintComponent(BaseCell cell, TPaintable p, PaintSwingContext<TImage> paintContext);

   /** залить ячейку нужным цветом */
   protected abstract void paintComponentBackground(BaseCell cell, TPaintable p, PaintSwingContext<TImage> paintContext);

   /** draw image Flag or image Mine */
   protected abstract void paintImage(BaseCell cell, TPaintable p, PaintSwingContext<TImage> paintContext, TImage img);

}

File: src/src_platform_uwp/fmg/uwp/draw/mosaic/CellPaint.cs
using fmg.core.mosaic.draw;
using fmg.core.mosaic.cells;

namespace fmg.uwp.draw.mosaic {

   /// <summary> Class for drawing cell </summary>
   /// <typeparam name="TPaintable">see <see cref="IPaintable"/></typeparam>
   /// <typeparam name="TImage">UWP specific image: <see cref="Windows.UI.Xaml.Media.Imaging.WriteableBitmap"/> or <see cref="Windows.UI.Xaml.Media.ImageSource"/>, etc... </typeparam>
   public abstract class CellPaint<TPaintable, TImage, TPaintContext> : ICellPaint<TPaintable, TImage, TPaintContext>
      where TPaintable : IPaintable
      where TImage : class
      where TPaintContext : PaintUwpContext<TImage>
   {

      public abstract void Paint(BaseCell cell, TPaintable paint, TPaintContext paintContext);

      protected abstract void PaintBorder(BaseCell cell, TPaintable paint, TPaintContext paintContext);

      /// <summary> draw border lines </summary>
      protected abstract void PaintBorderLines(BaseCell cell, TPaintable paint, TPaintContext paintContext);

      protected abstract void PaintComponent(BaseCell cell, TPaintable paint, TPaintContext paintContext);

      /// <summary> залить ячейку нужным цветом </summary>
      protected abstract void PaintComponentBackground(BaseCell cell, TPaintable paint, TPaintContext paintContext);

      /// <summary> draw image Flag or image Mine </summary>
      protected abstract void PaintImage(BaseCell cell, TPaintable paint, TPaintContext paintContext, TImage img);

   }

}
